File: May-LeetCoding-Challenge/13-Remove-K-Digits/Remove-K-Digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        string ans;
        for (auto& c : num){
            while (k and ans.length() and c < ans[ans.length() - 1]) {
                ans.pop_back();
                k--;
            }
            if (ans.length() == 0 and c == '0') continue;
            ans += c;
        }
         
        while (k-- and ans.length()) ans.pop_back();
        
        return ans.length()?ans:"0";
        
    }
};

File: May-LeetCoding-Challenge/13-Remove-K-Digits/Remove-K-Digits.scala
object Solution {
  def removeKdigits(num: String, k: Int): String =
    if (num.length == k) "0"
    else trimLeftZero(removeKdigitsWithAcc(num, num.length - k, ""))

  def removeKdigitsWithAcc(num: String, restNumberLength: Int, acc: String): String = {
    if (restNumberLength == 0) acc
    else if (num.length == restNumberLength) acc + num
    else {
      val rightIndexEnd = num.length - restNumberLength
      var smallestNumberIndex = 0
      var smallestNumber = num.head
      var index = 1
      while (index <= rightIndexEnd && smallestNumber != '0') {
        if (num(index).toInt < smallestNumber.toInt) {
          smallestNumber = num(index)
          smallestNumberIndex = index
        }
        index += 1
      }
      val newAcc = acc + smallestNumber
      removeKdigitsWithAcc(num.drop(smallestNumberIndex + 1), restNumberLength - 1, newAcc)
    }
  }

  def trimLeftZero(num: String): String = {
    val leftZeroDroppedNumber = num.dropWhile(_ == '0')
    if (leftZeroDroppedNumber.isEmpty) "0" else leftZeroDroppedNumber
  }

}
